[Blog post] Distributed Game of Life in Elixir

ArturT opened the pull request:
This is the first version of the article. 

TODO:
- [x] review chapters, make sense? etc
- [x] fix typos and ponglish/english 
- [x] update article release date to 9 May before we publish article
  `date:   2016-05-09 08:00:00 +0200`

Intro
- Demo
- Game of Life rules
- Create new application in Elixir
- Represent the board in Game of Life
- Game of Life rules with tests
- Architecture of distributed Game of Life
- Create BoardServer
- Create node manager for task supervisor
- Create board helper functions
- Create game printer and console presenter
- Add figure patterns and place them on the board
- Run game across multiple nodes

ArturT committed:
Initial post about game of life in elixir

ArturT committed:
Add intro

ArturT committed:
no quote

ArturT committed:
Create new application in Elixir section

ArturT committed:
Add table styles

ArturT committed:
Represent the board in Game of Life

ArturT committed:
Add images to show board with cells

ArturT committed:
Rules when cell should be alive or dead

ArturT committed:
Last requirement about dead cell

ArturT committed:
dead neighbours

ArturT committed:
Add arity to functions

ArturT committed:
gen server

ArturT committed:
add supervisor img

ArturT committed:
Describe architecture of game

ArturT committed:
Add section about board server

ArturT committed:
Create node manager for task supervisor

ArturT committed:
Fix typos

ArturT committed:
Add link to source code

ArturT committed:
remove task

ArturT committed:
Create board helper functions

ArturT committed:
Add todo headers

ArturT committed:
Add demo

ArturT committed:
Preload demo

ArturT committed:
Add logo

ArturT committed:
Refresh photos

ArturT committed:
intro

ArturT committed:
Fix problem with image visible in blog excerpt

ArturT committed:
no intro

ArturT committed:
game printer

ArturT committed:
Add figure patterns and place them on the board

ArturT committed:
Run game across multiple nodes

ArturT committed:
Add summary

ArturT committed:
lang fixes

ArturT committed:
Update what I'm going to show you is

ArturT committed:
Update demo section

ArturT committed:
link to next header

ArturT committed:
fixes

ArturT committed:
some text changes

ArturT committed:
update text

ArturT committed:
test changes

ArturT committed:
typo

ArturT committed:
fit code on the screen

ArturT committed:
typo

ArturT committed:
Add info about pattern rectangle position

ArturT committed:
small text changes

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+tags: elixir distributed game
+---
+
+I wrote my first game in Elixir. It's a popular thing - Conway's Game of Life - but it gets quite interesting when you solve it in a functional language, especially when you can see how [actor model](https://en.wikipedia.org/wiki/Actor_model) works and how actors are distributed across servers in network.
```
Avoid "it's". Try to use it is. Additionally, maybe we should make a link the sentence "how actor model works"?

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+<img src="/images/blog/posts/distributed-game-of-life-in-elixir/game_of_life_logo.png" style="float:right;margin-left:20px;" />
+
+__What I'm going to show you is:__
```
In this blog post I am going to show:

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+__What I'm going to show you is:__
+
+* How to write rules for game of life with tests in Elixir
+* Parallel tasks across lightweight processes (actors) so we utilize all CPU cores
```
so we utilize -> in order to utilize (?)

Tomabana replied:
"in order to utilize" sounds better

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+* How to distribute work across nodes so the game can be executed by many servers in the cluster
+* How to use GenServer behaviour, TaskSupervisor and Agents in Elixir
+
+This project and the full [source code can be found here](https://github.com/BeyondScheme/elixir-game_of_life).
```
Maybe we should mention in here that project can be found on official Beyond Scheme repository?

ArturT replied:
Any idea how this sentence could look like?

Tomabana replied:
This project and the full source code can be found [on the repository of our organization]. Maybe something like this?

ArturT replied:
@Tomabana Sounds too official.

This project and the full [source code can be found on our github]

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+<script type="text/javascript" src="https://asciinema.org/a/44233.js" id="asciicast-44233" async data-preload="true"></script>
+
+As you see we started node1 where the game is running and board is printed on the screen. We started also the second node and connected it to the first one. From the second node, we added new cells to the board. Both nodes are responsible for processing the game but only the first node is a master with info about the current state of the game. We can connect more nodes to the cluster so game proc ...
```
- "As you can see, node1 represents running game and board on the screen".
- "The second node was also started and connected to the first one".
- "Both nodes are responsible for processing the game," - comma after game
- "master with info" - with information

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+* Any live cell with more than three live neighbours dies, as if by over-population.
+* Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+
+The initial pattern constitutes the seed of the system. The first generation is created by applying the above rules simultaneously to every cell in the seed—births and deaths occur simultaneously, and the discrete moment at which this happens is sometimes called a tick (in other words, each generation is a pure function of the preceding one). The rules continue to be applied repeatedly to create  ...
```
"seed-births" - maybe consider to divide it into two sequences.

ArturT replied:
BTW this text is from Wikipedia. I was wondering about adding it in the quote but it looks bad when you have a wall of text as the quote so instead, I added the link to the source on wiki in the fist sentence for this chapter.

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Create new application in Elixir
+
+First thing first so we are going to create a new Elixir OTP application with supervision tree. We will use supervisor for our game server, you will learn more about it a bit later.
```
first thing first, - comma

Tomabana replied:
First thing**s** first?

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+<img src="/images/blog/posts/distributed-game-of-life-in-elixir/board_cells_xy.jpg" />
+
+Now when we have idea how we are going to store our alive cells we can jump to write some code.
```
the idea/an idea

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+---
```
A lot of "we".

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+Often in Elixir there is a need to share or store state that must be accessed from different processes or by the same process at different points in time. The Agent module provides a basic server implementation that allows state to be retrieved and updated via a simple API.
+This is what we are going to do in `GameOfLife.GamePrinter` as we need only keep time reference to our timer interval.
+
+# Create BoardServer
```
What do you think about splitting this post into two parts at this point?

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+Now we can implement our `GameOfLife.Cell.keep_alive?/2` function.
```
Now,

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+As you can see we implemented private function `count_neighbours/4` responsible for counting neighbours. It will be helpful later.
```
see,
- later - maybe directly point when/where.

ArturT replied:
Ok. I added something like this.

It will be helpful to meet the next rule.

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+As you can see we implemented private function `count_neighbours/4` responsible for counting neighbours. It will be helpful later.
+
+There is one more requirement we forogot which is:
```
forgot about,

Tomabana replied:
we forgot about:

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+* Parallel tasks across lightweight processes (actors) so we utilize all CPU cores
+* How to distribute work across nodes so the game can be executed by many servers in the cluster
+* How to use GenServer behaviour, TaskSupervisor and Agents in Elixir
+
```
I think we should separate these sentences with coma at the end of the line, and with dot at a last sentence.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Game of Life rules
+
+If you haven't heard about [game of life problem](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) then here is basic concept. If you already know it, just jump to [next header](#create-new-application-in-elixir).
```
If you already know about game of life problem just jump to the next header. If not, in this chapter you can learn basic concept.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+If you haven't heard about [game of life problem](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) then here is basic concept. If you already know it, just jump to [next header](#create-new-application-in-elixir).
+
+The universe of the Game of Life is an infinite two-dimensional orthogonal grid of square cells, each of which is in one of two possible states, alive or dead. Every cell interacts with its eight neighbours, which are the cells that are horizontally, vertically, or diagonally adjacent. At each step in time, the following transitions occur:
```
each of them?

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+alive_cells = [{0, 0}, {1, 0}, {2, 0}, {1, 1}, {-1,-2}]
+{% endhighlight %}
+
+Here is example how this board with alive cells looks like:
```
Here is an example.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+Let's write some tests to cover first of the requirement of the Game of Life.
```
of the Game of Life**:**

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+> Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+
+We are going to write a new function `GameOfLife.Cell.become_alive?/2` expecting coordinates of dead cell and returning if the dead cell should become alive or not.
```
a new function GameOfLife.Cell.become_alive?/2 which is expecting

ArturT replied:
grammarly.com plugin shows me that there is one advanced issue in the sentence you proposed but I don't have pro plan to see what that is so I'm going to leave the first version.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+defp neighbours([], neighbours), do: neighbours
+{% endhighlight %}
+
+Basically, those are all rules implemented in the single module `GameOfLife.Cell`. You can see the whole [module file](https://github.com/BeyondScheme/elixir-game_of_life/blob/master/lib/game_of_life/cell.ex) with [tests on GitHub](https://github.com/BeyondScheme/elixir-game_of_life/blob/master/test/game_of_life/cell_test.exs).
```
Basically, these ?

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+* `Task.Supervisor` is Elixir module defining a new supervisor which can be used to dynamically supervise tasks. We are going to use it to spin off tasks like determining if the particular cell should live or die. Those tasks can be run across nodes connected into the cluster. In above code, we gave name `GameOfLife.TaskSupervisor` for our supervisor. We will use this name to tell `Task.Superviso ...
+
+* `GameOfLife.BoardServer` is our module implemented as [GenServer behaviour](http://elixir-lang.org/docs/stable/elixir/GenServer.html). It's responsible for holding the state of the game. By that I mean it keeps the list of alive cells on the board along with generation counter and TRef. TRef is a timer reference returned by [Erlang timer module](http://erlang.org/doc/man/timer.html) and [apply_ ...
```
With each new generation we will update generation counter, (without first coma)

ArturT replied:
grammarly plugin shows me that there should be a comma in this:
"With each new generation, we will update generation counter."

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+Agent, on the other hand, is much simpler solution than GenServer. Agents are a simple abstraction around state.
+Often in Elixir there is a need to share or store state that must be accessed from different processes or by the same process at different points in time. The Agent module provides a basic server implementation that allows state to be retrieved and updated via a simple API.
+This is what we are going to do in `GameOfLife.GamePrinter` as we need only keep time reference to our timer interval.
```
as we need only **to** keep

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+In our public interface, we have `alive_cells/1` function which returns the list of alive cells. Basically, it's the current state of the game (alive cells on the board). This function calls `GenServer` with registered `@name` and request `:alive_cells`. We need to implement `handle_call/3` function for this type of request (`:alive_cells`).
+
+There is another public function `generation_counter/1` which returns how many generations was already processed by board server.
```
how many generations were already

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+There is another public function `generation_counter/1` which returns how many generations was already processed by board server.
+
+The `state/1` function returns state that is held by our generic server. The state is represented as the tuple with 3 values like alive cells, TRef (time reference - we want to regenerate board every second) and generation counter. TRef is very internal thing for board server so we won't return this to outside world. That's why we will return just alive cells and generation counter. You will see  ...
```
return this to the outside world.

Tomabana replied:
You will see it later in the implementation

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+The last 2 functions are `stop_game/1` and `change_speed/1` which just restart the game and starts a new one with provided speed.
+
+Now you can take a look how above functions are working exactly because they are calling server callbacks.
```
functions are working. (I think you can remove the rest of this sentence, or divide it to two separete sentences)

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+Oh, we forgot about tests. In this case, we can use [DocTest](http://elixir-lang.org/docs/stable/ex_unit/ExUnit.DocTest.html). It allows us to generate tests from the code examples existing in a module/function/macro’s documentation.
+
+Our test file is super short.
```
is super short**:**

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+Let's add `@moduledoc` do `GameOfLife.BoardServer`.
```
to `GameOfLife.BoardServer`**:**

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+We will add some useful functions to `GameOfLife.Board` module which helps us to do operations on the list of alive cells.
+
+Another thing you noticed is how we use `Task.Supervisor` in:
```
Another intersting thing is how we use

ArturT replied:
I guess another thing suits better in the context what was shown above.

ArturT replied:
my bad. I was looking at wrong sentence. I fixed that.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Create board helper functions
+
+We need some helper functions for operations we can do on board like adding, removing cells. Let's start with tests for module `GameOfLife.Board` and function `add_cells/2`.
```
we can do on the board

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+We need to ensure we won't allow to add the same cell twice to the board hence the above test. Here is implementation for `add_cells/2` function:
```
Here is the implementation

Tomabana replied:
to the board hence the above test? (What do you mean?)

ArturT replied:
Maybe this will be better?

We need to ensure we won’t allow adding the same cell twice to the board hence we test there are no duplicates.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+Another thing is removing cells from the list of alive cells.
```
of alive cells**:**

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Create game printer and console presenter
+
+`GameOfLife.GamePrinter` module is running as a worker under supervise of `GameOfLife` supervisor. `GameOfLife.GamePrinter` is using `Agent` to store `TRef` for timer reference as we want to print the board to the STDOUT with the specified interval. You already saw the example with using `Agent` so this shouldn't be new for you. Basically, we wrote public interface to start and stop printing the  ...
```
You have already saw the example with using

ArturT replied:
"You have already seen the example with using" would be correct.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+`GameOfLife.Presenters.Console` is responsible for printing board nicely with  X & Y axises, the number of alive cells and the generation counter.
```
is responsible for printing board with (without 'nicely')

ArturT replied:
It seems to be fine to use nicely in the context of printing.

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+Here is implemented our print function.
```
Here is implemented our print function**:**

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Add figure patterns and place them on the board
+
+To play our game of life would be great to have an easy way to add figures on the board. There are many common known patterns in game of life like still lifes, oscillators, spaceships. You can [learn more about them here](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life#Examples_of_patterns).
```
There are many common known patterns like still lifes

Tomabana commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Summary
+
+Finnally we managed to get to the end. It was a pretty long road but we have a working game, distributed across nodes. We learned how to write GenServer, how to use Agents and how we can split processes across nodes with TaskSupervisor. You also saw examples of tests in Elixir and how to use DocTest. Now you know how to connect nodes into the cluster.
```
We learned how to write GenServer,  use Agents. split processes across nodes with TaskSupervisor and connect nodes into the cluster. You also saw examples of tests in Elixir and how to use DocTest.

ArturT committed:
use it is instead of it's

ArturT committed:
link how actor model works

ArturT committed:
In this blog post I am going to show

ArturT committed:
in order to utilize

ArturT committed:
separate these sentences with coma at the end of the line, and with dot at a last sentence

ArturT committed:
Fixes in text under demo vid

ArturT committed:
change game of life rules intro text

ArturT committed:
missing words

ArturT committed:
First things first

ArturT committed:
Here is an example how this board with alive cells looks like:

ArturT committed:
the idea

ArturT committed:
colon

ArturT committed:
Now comma

ArturT committed:
comma

ArturT committed:
It will be helpful to meet the next rule.

ArturT committed:
forgot about

ArturT committed:
these

ArturT committed:
only need to

ArturT committed:
were

ArturT committed:
return this to the outside world

ArturT committed:
the

ArturT committed:
divide sentence

ArturT committed:
colon

ArturT committed:
to

ArturT committed:
the

ArturT committed:
the

ArturT committed:
Another intersting thing is how we use

ArturT committed:
colon

ArturT committed:
You have already seen the example

ArturT committed:
of

ArturT committed:
colon

ArturT committed:
remove repeating game of life

ArturT committed:
fixes in summary

ArturT committed:
We need to ensure we won’t allow adding the same cell twice to the board hence we test there are no duplicates.

ArturT committed:
Merge branch 'master' into game-of-life-blog-post

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+<img src="/images/blog/posts/distributed-game-of-life-in-elixir/supervisor.jpg" />
+
+Our main supervisor is `GameOfLife.Supervisor` which I mentioned at the begining of the article. Here you can see how we defined its childrens like `Task.Supervisor`, workers for `BoardServer` and `GamePrinter`.
```
Below you can see
children zamiast childrens

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+Let me describe you what each component on the image is responsible for.
```
let me describe - not need to use "describe you"

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+A GenServer is a process like any other Elixir process and it can be used to keep state, execute code asynchronously and so on. The advantage of using a generic server process (GenServer) is that it will have a standard set of interface functions and include functionality for tracing and error reporting. It also fits into a supervision tree as this is what we did in `GameOfLife` module.
+
+Agent, on the other hand, is much simpler solution than GenServer. Agents are a simple abstraction around state.
```
On the other hand, agent is

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+As you can see we use `GenServer` behaviour in our module. I require also Logger as we would like to print some info to the STDOUT.
```
As you can see,
Next sentence: Should it be "It requires"?

ArturT replied:
I will use "The module requires also Logger" to be more explicit.

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+As you can see we use `GenServer` behaviour in our module. I require also Logger as we would like to print some info to the STDOUT.
+
+In `start_link/1` function we start new `GenServer`. When our generic server was started as a first process in the cluster then it becomes master process. In the case when there is already running process with globally registered name `{:global, __MODULE__}` we log info that our process will be a slave process with a reference to existing PID on another node in the cluster.
```
was started? maybe we should use just "starts".
info -> information - or maybe you mean the log level. If yes, then leave it as it is.

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+In `start_link/1` function we start new `GenServer`. When our generic server was started as a first process in the cluster then it becomes master process. In the case when there is already running process with globally registered name `{:global, __MODULE__}` we log info that our process will be a slave process with a reference to existing PID on another node in the cluster.
+
+As you see we store global name for our server in attribute `@name`. We use another attribute `@game_speed` for default game speed which is 1000 miliseconds.
```
We store - without as you see.

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+end
+{% endhighlight %}
+
+This is the way how you can add the gosper glider pattern to the board with specified position.
```
without how.

natif commented on _posts/2016-04-26-distributed-game-of-life-in-elixir.md:
```diff
@@ -0,0 +1,1065 @@
+
+# Summary
+
+Finnally we managed to get to the end. It was a pretty long road but we have a working game, distributed across nodes. We learned how to write GenServer, use Agents, split processes across nodes with TaskSupervisor and connect nodes into the cluster. You also saw examples of tests in Elixir and how to use DocTest.
```
finally

ArturT committed:
children

ArturT committed:
Below you can see

ArturT committed:
Let me describe

ArturT committed:
On the other hand, agent

ArturT committed:
comma

ArturT committed:
more explicit who requires logger

ArturT committed:
starts

ArturT committed:
we store

ArturT committed:
just way

ArturT committed:
finally

ArturT committed:
Change permalink so post has url only with year instead of date when
the post file _posts/2016-04-26-distributed-game-of-life-in-elixir.md
was created. The publish date of post is likely later because post is
finished after a few days. We don't want to have incorrect date in url
when we publish post and we don't want to rename file every time when
we need to schedule post for a different date.

ArturT committed:
Add note how to show future dated posts

ArturT committed:
Set publish date to 2016-05-09 08:00:00 +0200

ArturT committed:
Remove category from blog post because we don't use category in permalink anymore

ArturT commented on README.md:
```diff
@@ -11,4 +11,8 @@ How to start server:
 
+You can write your blogs with future dated. Show future dated posts:
+
+    $ jekyll serve --future
```
I've set publish date for blog post to Monday so in order to see it you need to run project locally with jekyll serve --future

ArturT commented on _config.yml:
```diff
@@ -12,6 +12,7 @@ description: > # this means to ignore newlines until "baseurl:"
   We do Ruby on Rails and Java applications in Play! Framework or Spring MVC.
 baseurl: "" # the subpath of your site, e.g. /blog
 url: "http://beyondscheme.com" # the base hostname & protocol for your site
+permalink: /:year/:title
```
I've changed the permalink pattern to the simpler one. Thanks to that the post url won't have wrong date based on blog post file name where is date 2016-04-26 in _posts/2016-04-26-distributed-game-of-life-in-elixir.md

The old url had pattern /:categories/:year/:month/:day/:title.html
http://localhost:4000/elixir/2016/04/26/distributed-game-of-life-in-elixir.html

The new url:
http://localhost:4000/2016/distributed-game-of-life-in-elixir

ArturT replied:
You can remove your local directory `_site` where are old posts compiled with the old permalink pattern.
